### PR TITLE
Edge Drawing Improvements

### PR DESCRIPTION
**Pull request by sturkmen72**
#2314
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

[documentation preview](https://pullrequest.opencv.org/buildbot/export/pr_contrib/2907/docs/d4/d8b/group__ximgproc__edge__drawing.html)

```
force_builders=Linux x64,docs
```

### CODE REVIEW

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by sturkmen72** [comment]
@alalek please review

File: modules/ximgproc/src/edge_drawing.cpp [hunk 1]
```diff
@@ -251,6 +332,9 @@ void EdgeDrawingImpl::write(cv::FileStorage& fs) const
 EdgeDrawingImpl::EdgeDrawingImpl()
 {
     params = EdgeDrawing::Params();
+    nfa = new NFALUT(1, 1/2, 1, 1);
+    dH = new double[MAX_GRAD_VALUE];
+    grads = new int[MAX_GRAD_VALUE];
```
Comment (alalek) [inline modules/ximgproc/src/edge_drawing.cpp:337]: > new

Where are these objects destroyed?

---

> dH, grads

Consider using `std::vector` for these fields.

File: modules/ximgproc/src/edge_drawing.cpp [hunk 2]
```diff
@@ -337,6 +338,13 @@ EdgeDrawingImpl::EdgeDrawingImpl()
     grads = new int[MAX_GRAD_VALUE];
 }
 
+EdgeDrawingImpl::~EdgeDrawingImpl()
```
Comment (sturkmen72) [inline modules/ximgproc/src/edge_drawing.cpp:341]: is it enough for now ? i am planning to work on  optimizing all code later.

File: modules/ximgproc/perf/perf_edge_drawing.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+{
+/* 1. Define parameter type and test fixture */
+typedef tuple<int, int> EDTestParam;
+typedef TestBaseWithParam<EDTestParam> EdgeDrawing_Detect_Edges;
+typedef TestBaseWithParam<EDTestParam> EdgeDrawing_Detect_Lines;
+typedef TestBaseWithParam<EDTestParam> EdgeDrawing_Detect_Ellipses;
+
+/* 2. Declare the testsuite */
+PERF_TEST_P(EdgeDrawing_Detect_Edges, perf,
+            Combine(Values(0, 1, 2, 3), Values(10, 20, 30, 50, 100, 150, 200, 250)))
```
Comment (alalek) [inline modules/ximgproc/perf/perf_edge_drawing.cpp:19]: Do we really need to track performance of all these cases? Could we use subset of them here (and below)?

You can create separate full set (e.g, for "development"), but lets keep it disabled by default (see `DISABLED_FULL, OCL_GaussianBlurFixture` from `imgproc`)

File: modules/ximgproc/perf/perf_edge_drawing.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+{
+    /* 3. Get actual test parameters */
+    EDTestParam params = GetParam();
+    int EdgeDetectionOperator = get<0>(params);
+    int GradientThresholdValue = get<1>(params);
+
+    /* 4. Allocate and initialize arguments for tested function */
+    std::string filename = getDataPath("perf/1680x1050.png");
+    Mat src = imread(filename, 0);
```
Comment (alalek) [inline modules/ximgproc/perf/perf_edge_drawing.cpp:28]: > 0

Please avoid using of magic numbers. Use `IMREAD_GRAYSCALE` instead.

File: modules/ximgproc/src/edge_drawing.cpp [hunk 5]
```diff
@@ -12,6 +12,87 @@ namespace cv
...
+        uchar* dirRow = dirImage[y];
+
+        for (int x = 1; x < last_col; ++x)
+        {
+            int com1 = srcNextRow[x + 1] - srcPrevRow[x - 1];
+            int com2 = srcPrevRow[x + 1] - srcNextRow[x - 1];
+            int gx = 0;
+            int gy = 0;
+
+            switch (op)
```
Comment (alalek) [inline modules/ximgproc/src/edge_drawing.cpp:49]: BTW, "constant" mode switch in inner body of two loops may cause some performance loss.

(in the current patch it is just code move, so we can keep it "as is" in this PR)

File: modules/ximgproc/src/edge_drawing.cpp [hunk 6]
```diff
@@ -12,6 +12,87 @@ namespace cv
...
+                com2 = srcCurRow[x + 1] - srcNextRow[x];
+
+                gx = abs(com1 + com2);
+                gy = abs(com1 - com2);
+                break;
+            }
+
+            int sum;
+
+            if (SumFlag)
```
Comment (alalek) [inline modules/ximgproc/src/edge_drawing.cpp:75]: > if

similar to "constant" mode switch.

**Review by alalek — COMMENTED** [review]
Thank you for updates!

Could you please reduce amount of default performance test cases?

File: modules/ximgproc/perf/perf_edge_drawing.cpp (hunk 3 above)
Comment (sturkmen72) [reply modules/ximgproc/perf/perf_edge_drawing.cpp]: i will provide performans test later

**Review by alalek — APPROVED** [review]
Well done :+1:

**Pull request merged**